File: sample_java_project/GoodExample.java
package com.example.demo;

import java.util.List;
import java.util.ArrayList;

/**
 * A well-documented and properly structured class
 * demonstrating good Java coding practices.
 *
 * @author Developer
 * @version 1.0
 */
public class GoodExample {
    // Private fields with proper encapsulation
    private String name;
    private int age;
    private List<String> hobbies;

    /**
     * Constructor with parameter validation
     * @param name The person's name
     * @param age The person's age
     */
    public GoodExample(String name, int age) {
        if (name == null || name.trim().isEmpty()) {
            throw new IllegalArgumentException("Name cannot be null or empty");
        }
        if (age < 0) {
            throw new IllegalArgumentException("Age cannot be negative");
        }
        this.name = name;
        this.age = age;
        this.hobbies = new ArrayList<>();
    }

    /** Get the person's name */
    public String getName() { return name; }

    /** Set the person's name with validation */
    public void setName(String name) {
        if (name == null || name.trim().isEmpty()) {
            throw new IllegalArgumentException("Name cannot be null or empty");
        }
        this.name = name;
    }

    /** Get the person's age */
    public int getAge() { return age; }

    /** Set the person's age with validation */
    public void setAge(int age) {
        if (age < 0) {
            throw new IllegalArgumentException("Age cannot be negative");
        }
        this.age = age;
    }

    /** Add a hobby to the list */
    public void addHobby(String hobby) {
        if (hobby != null && !hobby.trim().isEmpty()) {
            hobbies.add(hobby);
        }
    }

    /** Get the list of hobbies */
    public List<String> getHobbies() { return new ArrayList<>(hobbies); }

    @Override
    public String toString() {
        return "GoodExample{name='" + name + "', age=" + age + ", hobbies=" + hobbies + "}";
    }
}

File: sample_cpp_project/src/GoodExample.cpp
#include "../include/GoodExample.h"
#include <stdexcept>
#include <sstream>
#include <algorithm>

GoodExample::GoodExample(const std::string& name, int age) {
    // Input validation with clear error messages
    if (name.empty()) {
        throw std::invalid_argument("Name cannot be empty");
    }
    if (age < 0) {
        throw std::invalid_argument("Age cannot be negative");
    }

    name_ = name;
    age_ = age;
    hobbies_.reserve(10); // Reserve reasonable capacity
}

const std::string& GoodExample::getName() const noexcept {
    return name_;
}

void GoodExample::setName(const std::string& name) {
    if (name.empty()) {
        throw std::invalid_argument("Name cannot be empty");
    }
    name_ = name;
}

int GoodExample::getAge() const noexcept {
    return age_;
}

void GoodExample::setAge(int age) {
    if (age < 0) {
        throw std::invalid_argument("Age cannot be negative");
    }
    age_ = age;
}

void GoodExample::addHobby(const std::string& hobby) {
    // Only add non-empty hobbies and avoid duplicates
    if (!hobby.empty()) {
        auto it = std::find(hobbies_.begin(), hobbies_.end(), hobby);
        if (it == hobbies_.end()) {
            hobbies_.push_back(hobby);
        }
    }
}

std::vector<std::string> GoodExample::getHobbies() const {
    // Return copy to maintain encapsulation
    return hobbies_;
}

std::string GoodExample::toString() const {
    std::ostringstream oss;
    oss << "GoodExample{name='" << name_ << "', age=" << age_ << ", hobbies=[";

    for (size_t i = 0; i < hobbies_.size(); ++i) {
        if (i > 0) {
            oss << ", ";
        }
        oss << "'" << hobbies_[i] << "'";
    }

    oss << "]}";
    return oss.str();
}